### ISSUE TITLE
Fixable mbox import issues ("from-munge" related)

### ISSUE THREAD

**Issue by ruffin--** [jstedfast/MimeKit#150]
I tried to use MimeKit to import some mboxes I took from Postbox 3.0.11, just to have some emails to play around with, and immediately ran into exceptions parsing a few email digests.

First, it's worth saying I realize these issues occur because Postbox doesn't seem to be following [From-munge rules](https://tools.ietf.org/html/rfc3676#section-4.4) properly. But I've seen Stedfast say he wants MimeKit to be ready for "real-world" use, so I'm reporting as an issue. I'm not sure if this is also a Thunderbird and potentially Postbox 4 issue, though they share code, as I have**n't** checked against either of those.

Here's the test code I'm using to read/import the messages, simplified a bit here:

    strMBox = Path.Combine(strMyProjectHome, "subset");
    mimeFormat = MimeFormat.Mbox;
	
	//...
	
    fs = File.OpenRead(strMBox);
    parser = new MimeParser(fs, mimeFormat);

    while (!parser.IsEndOfStream)
    {
        try
        {
            Console.WriteLine("##Attempting message: " + ++intMsgCnt + "##");
            var message = parser.ParseMessage();

            File.WriteAllText(Path.Combine(strEmlLoc, intMsgCnt + ".eml"), message.ToString());

            Console.WriteLine("##" + message.Subject + "##");
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
            Console.WriteLine(e.ToString());
            System.Diagnostics.Debugger.Break();
        }
    }

I've found two sorts of improper From-munges that break this code. The first is within a multipart message, which I think should be fixed (and is easy to do so), and another that's in a single-part text/plain digest email that's a little nastier.

I'm going to simplify each of the error cases, below, but I've placed the full source of two violating emails on Pastebin:

* [Multipart](http://pastebin.com/HWRU5pGm) (line 94)
* [Single-part](http://pastebin.com/6Fd6FDkE) (line 505)

The source at those links have their initial mbox "From " boundary lines, but are not in a real mbox context. That is, it's one email a piece there.

The issues are just as you'd expect. Postbox has lines *within the emails' bodies* that start with `From ` (with the trailing space, no leading `>` for quoting), so `IsMboxMarker` somewhat naively returns `true` for these lines even though they're not intended to be mbox markers.

**Multipart (with fix)**

I can see not from-munging in a multi-part message, since you have [a very rare pattern of bytes defining a boundary](https://github.com/jstedfast/MimeKit/blob/0cc8dae338a26c4cb4597b52a4dd27d3407d05b2/MimeKit/MimeParser.cs#L65) to say that you're in a multi-part proper, not in a email's headers. I think that's a valid issue for MimeKit. I haven't completely grokked all the MimeKit code, but MimeKit does seem in my limited testing to call `PopBoundary` properly at the conclusion of an email (again, end boundaries for multipart emails are very distinct, so you'd expect reliable handling).

This means we can fix "missing" from-munges in multipart emails with a very simple check in MimeParser's `IsPossibleBoundary` ([currently line 1066](https://github.com/jstedfast/MimeKit/blob/0cc8dae338a26c4cb4597b52a4dd27d3407d05b2/MimeKit/MimeParser.cs#L1066)):

Change from:  
`if (format == MimeFormat.Mbox && length >= 5 && IsMboxMarker (text))`

To:
`if (format == MimeFormat.Mbox && length >= 5 && IsMboxMarker (text) && 1 == this.bounds.Count)`

I don't think that's a huge performance issue -- no degradation at all if you're not parsing an mbox, and if you put the `Count` check before `IsMboxMarker`, it could be faster in many use cases too, I bet -- and that change seems properly defensive for a "live" parser. 

I don't believe there's a use case for having an entry in `bounds` other than "From " *outside* of an email, but I may have missed something.

**Single-part**

The single-part email is more troublesome. You could catch most issues by checking for at least one valid header in the line immediately after the possible mbox marker/"From " line, [checking for header validity similiar to what's happening in `StepHeaders` now](https://github.com/jstedfast/MimeKit/blob/0cc8dae338a26c4cb4597b52a4dd27d3407d05b2/MimeKit/MimeParser.cs#L880). That wouldn't completely solve this issue if you had the misfortune of a faux-header afterwards, but you've already *greatly* reduced your edge case failures. 

That is, this would still bork:

    From a library in Jersey, I checked out the book
    Mysteries: An mbox excursion.

It doesn't seem unfair to have MimeKit users preprocess mbox files before import to perform this sort of check for missing single-part email from-munges. The header check (which would still fail on the above) would complicate `IsPossibleBoundary` just the other side of nontrivially. That said, I'll probably add something for my local copy anyhow if MimeKit doesn't. The header check isn't a huge burden for the added defense. Though perhaps you could have a headerless email? Not sure if that's valid MIME without a Content-Type though.

I think the lesson is, of course, that the "standard that isn't" called mbox isn't particularly robust. Kind of maddening, actually. It'd've been easy to have made better mbox message boundaries than "From [something]". I mean, heck, multipart messages already have good boundary markers. Painful. 

Course if Postbox had from-munged more carefully, this wouldn't be a problem either. Though since Postbox doesn't, I'd have to think Thunderbird has the same issue -- hard to believe they'd go through the trouble to introduce a much less reliable mbox format. /shrug

<hr>

**Error cases**

Again, full source with minimal changes can be found at pastebin:

* [Multipart](http://pastebin.com/HWRU5pGm) (line 94)
* [Single-part](http://pastebin.com/6Fd6FDkE) (line 505)

What follows are cut-down versions of those just for quick reference. Apologies if I'm missing something here.

Multipart example:

	From - Sun Jun 21 21:30:06 2015
	Received: by 10.140.19.233 with SMTP id 96csp110646qgh;
	        Wed, 20 May 2015 22:26:28 -0700 (PDT)
	Return-Path: <mozilla.dev.apps.thunderbird+bncBCC6L7NBYACBBA6Z6WVAKGQET33QIWA@googlegroups.com>
	MIME-Version: 1.0
	To: Digest recipients <mozilla.dev.apps.thunderbird@googlegroups.com>
	From: mozilla.dev.apps.thunderbird@googlegroups.com
	Subject: Digest for mozilla.dev.apps.thunderbird@googlegroups.com - 2 updates in 2 topics
	Message-ID: <90e6ba6144440170ff051690c91f@google.com>
	Date: Thu, 21 May 2015 05:26:26 +0000
	Content-Type: multipart/alternative; boundary=0016367f94b4b30060474d22062
	Reply-To: mozilla.dev.apps.thunderbird@googlegroups.com

	--0016367f94b4b30060474d22062
	Content-Type: text/plain; charset=UTF-8

	---------- 1 of 1 ----------
	From: "ISHIKAWA, Chiaki" <redactd@yk.rim.or.jp>
	Date: May 21 09:55AM +0900
	Url: http://groups.google.com/group/mozilla.dev.apps.thunderbird/msg/2da08bdeca9b922c

	During local |make mozmill| test using full DEBUG version of TB, I noticed
	the following error in the log.

	From a local summary of errors/warnings:

	     11 JavaScript error: chrome://messenger/content/folderDisplay.js,
	line 2480: TypeError: this.view.dbView is null

	11 at the beginning, this happened 11 (eleven) times.

	I finally have had the time to look at this issue. The first and the
	second appearance of this error are as follows. This doesn't look
	right.  How can a test pass where we have such glaring errors?!

	--0016367f94b4b30060474d22062
	Content-Type: text/html; charset=UTF-8
	  <table style="border-collapse: collapse; width: 100%">
	    
	      <tr><td style="background-color: #FFFFFF; color:#2E2E2E; font-family: arial; padding:10px 15px; border:1px solid #d3d3d3;">
	        <span style="color:#B1B0B0; font-size: 15px;">
	          &quot;ISHIKAWA, Chiaki&quot; &lt;redactd@yk.rim.or.jp&gt;: May 21 09:55AM +0900
	        </span>
	        <br /><br />
	        During local |make mozmill| test using full DEBUG version of TB, I noticed<br />
	the following error in the log.<br />
	&nbsp;<br />
	From a local summary of errors/warnings:<br />
	&nbsp;<br />
	     11 JavaScript error: chrome://messenger/content/folderDisplay.js,<br />
	line 2480: TypeError: this.view.dbView is null<br />
	&nbsp;<br />
	11 at the beginning, this happened 11 (eleven) times.<br />
	&nbsp;<br />
	I finally have had the time to look at this issue. The first and the<br />
	second appearance of this error are as follows. This doesn&#39;t look<br />
	right.  How can a test pass where we have such glaring errors?!<br />
	&nbsp;<br />
	
	--0016367f94b4b30060474d22062--

<hr>

Single-part example:

	From - Sun Jun 21 21:30:11 2015
	Delivered-To: redactd@gmail.com
	Received: by 10.140.100.165 with SMTP id s34csp633627qge;
	        Wed, 27 May 2015 22:00:11 -0700 (PDT)
	Return-Path: <owner-LIBLICENSE-L@listserv.crl.edu>
	Date: Thu, 28 May 2015 00:00:00 -0500
	Reply-To: LibLicense-L Discussion Forum <LIBLICENSE-L@LISTSERV.CRL.EDU>
	Sender: LibLicense-L Discussion Forum <LIBLICENSE-L@LISTSERV.CRL.EDU>
	From: LIBLICENSE-L automatic digest system <LISTSERV@LISTSERV.CRL.EDU>
	Subject: LIBLICENSE-L Digest - 26 May 2015 to 27 May 2015 (#2015-98)
	To: <LIBLICENSE-L@LISTSERV.CRL.EDU>
	Message-ID: <LIBLICENSE-L%201505280000003140.16CB@LISTSERV.CRL.EDU>
	Content-Type: text/plain; charset="UTF-8"
	Content-Transfer-Encoding: quoted-printable
	
	There are 5 messages totaling 701 lines in this issue.

	Topics of the day:

	  5. June 17 NISO Virtual Conference: The Eternal To-Do List: Making Ebooks
	     work in Libraries

	------------------------------

	Date:    Wed, 27 May 2015 18:46:33 -0400
	From:    LIBLICENSE <redactd@GMAIL.COM>
	Subject: June 17 NISO Virtual Conference: The Eternal To-Do List: Making Ebooks work in Libraries

	From: NISO Announce <redactd@niso.org>
	Date: Wed, 27 May 2015 11:45:37 -0400

	June 17 NISO Virtual Conference: The Eternal To-Do List: Making Ebooks
	work in Libraries

	Virtual conferences are 5-6 hour conferences held online in
	webinar-like formats, with occasional breaks in the schedule for
	participants. The longer length allows the depth of coverage of a
	conference coupled with the convenience of a webinar.

	Date: June 17, 2015

	Time: 11:00 am - 5:00 pm Eastern

	Event webpage: http://www.niso.org/news/events/2015/virtual_conferences/ete=
	rnal_ebooks/

	ABOUT THE VIRTUAL CONFERENCE

	From scholarly monographs to textbooks, the range of e-book formats
	and use cases is rapidly expanding. Libraries are on the front lines
	of this issue, actively evolving their approach to offering e-books to
	meet patron needs and expectations. This virtual conference, The
	Eternal To-Do List: Making E-books Work in Libraries, will probe the
	key issues surrounding e-books from a variety of library, technology,
	and end-user viewpoints and share experiences of how some libraries
	have met these challenges.

	End of LIBLICENSE-L Digest - 26 May 2015 to 27 May 2015 (#2015-98)
	******************************************************************

To be clear, the problem in the second one starts at "From scholarly monographs..."

Also, this is the opposite side of the problem discussed [here](https://github.com/jstedfast/MimeKit/issues/108). That issue was with from-munges not having their munge stripped. This deals with mboxes that are *not* from-munged properly.

Thanks for the excellent libraries!

**Comment by jstedfast** [comment]
The Multipart issue is impossible to fix. Your work-around idea won't work because it's not necessary for end boundaries to be present.

The Single-part example is not worth fixing either.

I would just recommend avoiding Postbox and/or file bugs against their software to get them to fix these issues.

If they aren't going to properly munge From-lines and not even use the Content-Length alternative "standard" (which MimeKit supports), they are completely and utterly broken.

FWIW, I don't think Thunderbird has this issue as I use it and have never run into this.

**Comment by jstedfast** [comment]
FWIW, I've just confirmed that Thunderbird properly munges From-lines.

**Comment by ruffin--** [comment]
Wonder what Postbox was thinking.  I'll tell them, though perhaps v4 already has that fixed. This is on OS X, fwiw.

It [looks like boundary delimiters are required](http://tools.ietf.org/html/rfc2046#section-5.1.1); just out of curiousity, what system commonly omits them?

It looks like MIME requires at least `MIME-Version` in headers. If so, the header check for the line after a possible mbox boundary seems worth doing, even if I do it in a preprocessing step. That feels appropriately defensive.

Wonder if it's worth parsing from the end of the mbox instead of from the top. Seems that plus header-check is more robust for mbox than what MimeKit is doing now, though if it's just Postbox (or maybe even just Postbox 3) that's commonly screwing from-munging up, I'd agree that's pretty edgy.

Thanks for the follow-up and for checking Thunderbird.

**Comment by jstedfast** [comment]
> It looks like boundary delimiters are required; just out of curiousity, what system commonly omits them?

One of the biggest problems that most mime parser libraries suffer from is their reliance on the notion that when the spec says something is required, that other implementations will actually follow the spec completely.

(Note: the specs also say that headers must never contain non-ascii characters, yet I'm sure you've received email that do - my mailboxes are full of real-world examples).

In my experience with real-world messages, there are instances where the end boundaries have not existed. It may just be that there was inadvertent truncation of the message at some point during transmission, but with all of the breakage I've seen, it would not surprise me in the least if certain MIME sending implementations just didn't write them out in the first place.

> It looks like MIME requires at least MIME-Version in headers.

`MIME-Version` is not actually a required header unless you use one of the Content-* headers. Remember: MIME is an *optional* extension to the [Internet Message Format](https://tools.ietf.org/html/rfc5322). That said, the IMF spec also specifies that certain headers are required... (I believe those are `Date` and `From` from memory), so checking for something that looks like a header after the From-line is a possibility, but so far the only mail software that this is needed for seems to be Postbox, so it hardly seems worth the effort.

After testing this in Thunderbird last night and looking through Postbox's Release Notes for a fix, I noticed that there was a fix in 3.0.8(?) that they fixed importing from Thunderbird - that seemed very odd to me if they are using the same code. But this got me thinking that they probably are not using mbox at all, but rather some custom format that sorta-kinda-looks-like mbox. They might have decided in their infinite wisdom to create their own "mbox" format that likely uses some sort of index or database file alongside the mbox that records message start and end offsets so that they don't have to munge From-lines at all.

For what it's worth, I just tested how Thunderbird handles unmunged from lines using the following mbox:

```
From -
From: Example Sender <sender@example.com>
Subject: This is a test message with an unmunged From line
MIME-Version: 1.0
Content-Type: text/plain

This is my favorite James Bond movie:

From Russia with love

What does Thunderbird do?

From -
From: Example Sender <sender@example.com>
Subject: Another test message with an unmunged From line
MIME-Version: 1.0
Content-Type: text/plain

This is my favorite James Bond movie:

From Russia with love
Year: 1963

What did Thunderbird do this time?
```

The result in Thunderbird is 4 messages.
